Cabal/Distribution/Simple/GHC.hs

```diff
@@ -1283,6 +1321,34 @@ gbuild verbosity numJobs pkg_descr lbi bm clbi = do
     runGhcProg compileOpts { ghcOptNoLink  = toFlag True
                            , ghcOptNumJobs = numJobs }
 
+  -- build any C++ sources
+  unless (null cxxSrcs) $ do
+   info verbosity "Building C++ Sources..."
+   sequence_
+     [ do let baseCxxOpts    = Internal.componentCxxGhcOptions verbosity implInfo
+                               lbi bnfo clbi tmpDir filename
+              vanillaCxxOpts = if isGhcDynamic
+                                -- Dynamic GHC requires C++ sources to be built
+                                -- with -fPIC for REPL to work. See #2207.
+                               then baseCxxOpts { ghcOptFPic = toFlag True }
+                               else baseCxxOpts
+              profCxxOpts    = vanillaCxxOpts `mappend` mempty {
+                                 ghcOptProfilingMode = toFlag True
+                               }
+              sharedCxxOpts  = vanillaCxxOpts `mappend` mempty {
+                                 ghcOptFPic        = toFlag True,
+                                 ghcOptDynLinkMode = toFlag GhcDynamicOnly
+                               }
+              opts | needProfiling = profCxxOpts
+                   | needDynamic   = sharedCxxOpts
+                   | otherwise     = vanillaCxxOpts
+              odir = fromFlag (ghcOptObjDir opts)
```

Comment:
Possibly stupid question: the C++ and C files are sharing an output directory here, so collisions are a possibility. Is that user error if it happens, or do we need to guard against it?

Reply:
OK. Might be an idea to warn about it (either here or in `cabal check` or both), but that can be done later.

Reply:
I added a TODO for this potential collision issue. Should I create a GitHub issue for this so that it is visibly tracked.